File: construct-binary-tree-from-preorder-and-inorder-traversal/flynn.cpp
/**
 * For the number of given nodes N,
 * 
 * Time complexity: O(N)
 * 
 * Space complexity: O(N) at worst
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        unordered_map<int, int> inorder_index_map;
        stack<TreeNode*> tree_stack;

        for (int i = 0; i < inorder.size(); i++) inorder_index_map[inorder[i]] = i;

        TreeNode* root = new TreeNode(preorder[0]);
        tree_stack.push(root);

        for (int i = 1; i < preorder.size(); i++) {
            TreeNode* curr = new TreeNode(preorder[i]);

            if (inorder_index_map[curr->val] < inorder_index_map[tree_stack.top()->val]) {
                tree_stack.top()->left = curr;
            } else {
                TreeNode* parent;
                while (!tree_stack.empty() && inorder_index_map[curr->val] > inorder_index_map[tree_stack.top()->val]) {
                    parent = tree_stack.top();
                    tree_stack.pop();
                }
                parent->right = curr;
            }
            tree_stack.push(curr);
        }

        return root;
    }
};

File: same-tree/flynn.go
/*
풀이
- 재귀함수를 이용해서 풀이할 수 있습니다
Big O
- N: 트리 노드의 개수
- H: 트리의 높이 (logN <= H <= N)
- Time complexity: O(N)
  - 모든 노드를 최대 1번 탐색합니다
- Space complexity: O(H)
  - 재귀 호출 스택의 깊이는 H에 비례하여 증가합니다
*/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func isSameTree(p *TreeNode, q *TreeNode) bool {
	// base case
	if p == nil && q == nil {
		return true
	} else if p == nil || q == nil {
		return false
	}

	if p.Val != q.Val {
		return false
	}

	if !isSameTree(p.Left, q.Left) || !isSameTree(p.Right, q.Right) {
		return false
	}

	return true
}
